### PR TITLE
Cleanup GC *_STAT bitrot

### PR DESCRIPTION
**Pull request by sdmaclea**

### CODE REVIEW

**Comment by sdmaclea** [comment]
@Maoni0 @swgillespie These are instrumentation changes I made to JOIN_STATS & SYNCHRONIZATION_STATS to get them working on tip for ARM64.

I cleaned them up because I thought they would be generally helpful

File: src/gc/gc.cpp [hunk 1]
```diff
@@ -545,16 +547,18 @@ process_sync_log_stats()
...
-            "SuspendEE: %8u; RestartEE: %8u\n",
+            "SuspendEE: %8u; RestartEE: %8u GC%% %g\n",
             log_interval / 1000,
             gc_count_during_log,
             gc_lock_contended,
             (unsigned int)(gc_during_log / gc_count_during_log),
             (unsigned int)(suspend_ee_during_log / gc_count_during_log),
-            (unsigned int)(restart_ee_during_log / gc_count_during_log));
+            (unsigned int)(restart_ee_during_log / gc_count_during_log),
+            (double)(100.0f * gc_during_log / total));
```
Comment (sdmaclea) [inline src/gc/gc.cpp]: Added % time GC is running.

File: src/gc/gc.cpp [hunk 2]
```diff
@@ -899,39 +893,69 @@ class t_join
         }
     }
 
+#ifdef JOIN_STATS
+    uint64_t getTS()
+    {
+        return GCToOSInterface::QueryPerformanceCounter();
```
Comment (sdmaclea) [inline src/gc/gc.cpp]: GetCycleCount32() is AMD64 assemby.  Prefer GCToOSInterface::QueryPerformanceCounter().  Centralize by using getTS()

File: src/gc/gc.cpp [hunk 3]
```diff
@@ -882,12 +882,6 @@ class t_join
                 }
 
                 fire_event (gch->heap_number, time_end, type_join, join_id);
-
-#ifdef JOIN_STATS
-                // parallel execution starts here
-                start[gch->heap_number] = GetCycleCount32();
-                Interlocked::ExchangeAdd(&in_join_total[join_id], (start[gch->heap_number] - end[gch->heap_number])/1000);
-#endif //JOIN_STATS
```
Comment (sdmaclea) [inline src/gc/gc.cpp]: Reverse join is different and implementation was incomplete.  Stop collecting statistics instead of collecting bad statistics.

File: src/gc/gc.cpp [hunk 4]
```diff
@@ -899,39 +893,69 @@ class t_join
...
+
         // enable this printf to get statistics on each individual join as it occurs
-//      printf("join #%3d  seq_loss = %5d   par_loss = %5d  efficiency = %3.0f%%\n", join_id, seq_loss/1000, par_loss/1000, efficiency);
+//      printf("join #%3d  seq_loss = %5g   par_loss = %5g  efficiency = %3.0f%%\n", join_id, tsScale*seq_loss, tsScale*par_loss, efficiency);
 
-        elapsed_total[join_id] += sum/1000;
-        seq_loss_total[join_id] += seq_loss/1000;
-        par_loss_total[join_id] += par_loss/1000;
+        Interlocked::ExchangeAdd(&elapsed_total[id], sum);
+        Interlocked::ExchangeAdd(&wake_total[id], wake);
```
Comment (sdmaclea) [inline src/gc/gc.cpp]: wake_total is an estimate in the amount of time taken to wake all the threads.

File: src/gc/gc.cpp [hunk 5]
```diff
@@ -1657,12 +1682,12 @@ static void leave_spin_lock(GCSpinLock *pSpinLock)
 //the gc thread call WaitLonger.
 void WaitLonger (int i
 #ifdef SYNCHRONIZATION_STATS
-    , VOLATILE(GCSpinLock)* spin_lock
+    , GCSpinLock* spin_lock
```
Comment (sdmaclea) [inline src/gc/gc.cpp]: Code was not written to support VOLATILE() -> Volatile<T>

File: src/gc/gc.cpp [hunk 6]
```diff
@@ -13096,22 +13121,22 @@ int gc_heap::try_allocate_more_space (alloc_context* acontext, size_t size,
...
 #ifdef SYNCHRONIZATION_STATS
-    unsigned int msl_acquire_start = GetCycleCount32();
+    int64_t msl_acquire_start = GCToOSInterface::QueryPerformanceCounter();
 #endif //SYNCHRONIZATION_STATS
     enter_spin_lock (&more_space_lock);
     add_saved_spinlock_info (me_acquire, mt_try_alloc);
     dprintf (SPINLOCK_LOG, ("[%d]Emsl for alloc", heap_number));
 #ifdef SYNCHRONIZATION_STATS
-    unsigned int msl_acquire = GetCycleCount32() - msl_acquire_start;
+    int64_t msl_acquire = GCToOSInterface::QueryPerformanceCounter() - msl_acquire_start;
```
Comment (sdmaclea) [inline src/gc/gc.cpp]: CycleCount32() is AMD specific.  Avoid it.

File: src/gc/gc.cpp [hunk 7]
```diff
@@ -16522,6 +16547,9 @@ int gc_heap::garbage_collect (int n)
 
     fix_allocation_contexts (TRUE);
 #ifdef MULTIPLE_HEAPS
+#ifdef JOIN_STATS
+    gc_t_join.startTS(this);
+#endif //JOIN_STATS
```
Comment (sdmaclea) [inline src/gc/gc.cpp]: For the first join in a GC elpased time was incorrect unless this was added.

File: src/gc/gc.cpp [hunk 8]
```diff
@@ -545,16 +547,18 @@ process_sync_log_stats()
 
     if (log_elapsed > log_interval)
     {
+        uint64_t total = GCToOSInterface::QueryPerformanceCounter() - log_start_hires;
         // Print out the cycles we spent on average in each suspend and restart.
         printf("\n_________________________________________________________________________________\n"
             "Past %d(s): #%3d GCs; Total gc_lock contended: %8u; GC: %12u\n"
-            "SuspendEE: %8u; RestartEE: %8u\n",
+            "SuspendEE: %8u; RestartEE: %8u GC%% %g\n",
```
Comment (Maoni0) [inline src/gc/gc.cpp]: >GC%% %g [](start = 44, length = 7)

I would do GC %.3f%% here to be consistent with other percentages' format and .3 is sufficient.

File: src/gc/gc.cpp (hunk 1 above)
Comment (Maoni0) [inline src/gc/gc.cpp]: For the format I would do GC %.3f%% here to be consistent with other percentages' format and .3 is sufficient.

File: src/gc/gc.cpp (hunk 2 above)
Comment (Maoni0) [inline src/gc/gc.cpp]: please try to conform to GC's coding style - rename getTS to something like get_ts or get_current_time and starts accordingly.

File: src/gc/gc.cpp (hunk 4 above)
Comment (Maoni0) [inline src/gc/gc.cpp]: no need to change these to Interlocked - restart is only ever called on one thread.

File: src/gc/gc.cpp [hunk 9]
```diff
@@ -1657,12 +1682,12 @@ static void leave_spin_lock(GCSpinLock *pSpinLock)
...
 void WaitLonger (int i
 #ifdef SYNCHRONIZATION_STATS
-    , VOLATILE(GCSpinLock)* spin_lock
+    , GCSpinLock* spin_lock
 #endif //SYNCHRONIZATION_STATS
     )
 {
 #ifdef SYNCHRONIZATION_STATS
-    (spin_lock->num_wait_longer)++;
+    Interlocked::Increment(&spin_lock->num_wait_longer);
```
Comment (Maoni0) [inline src/gc/gc.cpp]: I am on the fence about these ++ -> Interlocked add changes. I didn't want to use Interlocked 'cause it's expensive and these counts don't need to be perfectly accurate. Have you measured the overhead with these Interlocked operations when you define the STATS? If you take a test that just does straight up allocations, how much overhead do you notice with these Interlocks?

File: src/gc/gc.cpp [hunk 10]
```diff
@@ -34934,7 +34962,7 @@ GCHeap::GarbageCollectGeneration (unsigned int gen, gc_reason reason)
         if (localCount != col_count)
         {
 #ifdef SYNCHRONIZATION_STATS
-            gc_lock_contended++;
+            Interlocked::Increment(&gc_lock_contended);
```
Comment (Maoni0) [inline src/gc/gc.cpp]: this is already under a lock so shouldn't be changed to interlocked.

File: src/gc/gc.cpp [hunk 11]
```diff
@@ -899,39 +893,69 @@ class t_join
...
         }
-        unsigned int seq_loss = (join_struct.n_threads - 1)*elapsed_seq;
-        unsigned int par_loss = join_struct.n_threads*max - sum;
+        uint64_t seq_loss = (join_struct.n_threads - 1)*elapsed_seq;
+        uint64_t par_loss = join_struct.n_threads*max - sum;
         double efficiency = 0.0;
         if (max > 0)
             efficiency = sum*100.0/(join_struct.n_threads*max);
 
+        const double tsScale = 1e-6;
```
Comment (Maoni0) [inline src/gc/gc.cpp]: tsScale -> ts_scale

File: src/gc/gc.cpp (hunk 10 above)
Comment (sdmaclea) [inline src/gc/gc.cpp]: I was using this as a shortcut for VolatileStore = VolatileLoad + 1.  even though there is no race, there is possibility of data loss as store buffer on one thread are not flushed.  But again sisnce thse are rare it is not really needed.

**Comment by sdmaclea** [comment]
@Maoni0 I believe this push resolved your comments.

**Comment by Maoni0** [comment]
LGTM. Thanks for doing this work!

**Comment by sdmaclea** [comment]
Ubuntu arm64 is broken in CI.  Please ignore and merge.
